worlds/sc2/rules.py

```diff
@@ -2518,76 +2533,131 @@ def end_game_requirement(self, state: CollectionState) -> bool:
+                    item_names.THORNSHELL, item_names.HUNTERLING,
+                ), self.player)
+                or state.has_all((item_names.INFESTOR, item_names.INFESTOR_INFESTED_TERRAN), self.player)
+                or self.morph_baneling(state)
+                or self.morph_lurker(state)
+                or self.morph_impaler(state)
+                or self.morph_brood_lord(state)
+                or self.morph_guardian(state)
+                or self.morph_ravager(state)
+                or self.morph_igniter(state)
+                or self.morph_tyrannozor(state)
+                or self.morph_devourer(state) and state.has(item_names.DEVOURER_PRESCIENT_SPORES, self.player)
+            )
+        return _has_zerg_units
+
+    def has_protoss_units(self, target: int) -> Callable[['CollectionState'], bool]:
+        def _has_protoss_units(state: CollectionState) -> bool:
+            return (
+                state.count_from_list_unique(
+                    item_groups.protoss_units
+                    + item_groups.protoss_buildings
+                    + [item_names.NEXUS_OVERCHARGE],
+                    self.player
+                ) >= target
+            ) and (
+                # Anything that can hit buildings
+                state.has_any((
+                    # Gateway
+                    item_names.ZEALOT, item_names.CENTURION, item_names.SENTINEL, item_names.SUPPLICANT,
+                    item_names.STALKER, item_names.INSTIGATOR, item_names.SLAYER,
+                    item_names.DRAGOON, item_names.ADEPT,
+                    item_names.SENTRY, item_names.ENERGIZER,
+                    item_names.AVENGER, item_names.DARK_TEMPLAR, item_names.BLOOD_HUNTER,
+                    item_names.HIGH_TEMPLAR, item_names.SIGNIFIER, item_names.ASCENDANT,
+                    item_names.DARK_ARCHON,
+                    # Robo
+                    item_names.IMMORTAL, item_names.ANNIHILATOR, item_names.VANGUARD, item_names.STALWART,
+                    item_names.COLOSSUS, item_names.WRATHWALKER,
+                    item_names.REAVER, item_names.DISRUPTOR,
+                    # Stargate
+                    item_names.SKIRMISHER,
+                    item_names.SCOUT, item_names.MISTWING, item_names.OPPRESSOR,
+                    item_names.INTERCESSOR, item_names.VOID_RAY, item_names.DESTROYER, item_names.DAWNBRINGER,
+                    item_names.ARBITER, item_names.ORACLE,
+                    item_names.CARRIER, item_names.TRIREME, item_names.SKYLORD,
+                    item_names.TEMPEST, item_names.MOTHERSHIP,
+                ), self.player)
+                or state.has_all((item_names.WARP_PRISM, item_names.WARP_PRISM_PHASE_BLASTER), self.player)
+                or state.has_all((item_names.CALADRIUS, item_names.CALADRIUS_CORONA_BEAM), self.player)
+                or state.has_all((item_names.PHOTON_CANNON, item_names.KHALAI_INGENUITY), self.player)
```

Comment:
Khalai Ingenuity is not progression item. Add a test for any units logic level (currently it scans only standard and advanced tactics)